File: internal/types/organization.go
package types

import (
	"slices"
	"time"

	"github.com/distr-sh/distr/internal/dbcrypto"
	"github.com/distr-sh/distr/internal/limit"
	"github.com/google/uuid"
)

type Organization struct {
	ID                                  uuid.UUID          `db:"id" json:"id"`
	CreatedAt                           time.Time          `db:"created_at" json:"createdAt"`
	Name                                string             `db:"name" json:"name"`
	Slug                                *string            `db:"slug" json:"slug"`
	Features                            []Feature          `db:"features" json:"features"`
	SubscriptionType                    SubscriptionType   `db:"subscription_type" json:"subscriptionType"`
	SubscriptionPeriod                  SubscriptionPeriod `db:"subscription_period" json:"subscriptionPeriod"`
	SubscriptionEndsAt                  time.Time          `db:"subscription_ends_at" json:"subscriptionEndsAt"`
	StripeCustomerID                    *string            `db:"stripe_customer_id" json:"stripeCustomerId"`
	StripeSubscriptionID                *string            `db:"stripe_subscription_id" json:"stripeSubscriptionId"`
	SubscriptionCustomerOrganizationQty limit.Limit        `db:"subscription_customer_organization_quantity" json:"subscriptionCustomerOrganizationQuantity"` //nolint:lll
	SubscriptionUserAccountQty          limit.Limit        `db:"subscription_user_account_quantity" json:"subscriptionUserAccountQuantity"`                   //nolint:lll
	PreConnectScript                    *string            `db:"pre_connect_script" json:"preConnectScript"`
	PostConnectScript                   *string            `db:"post_connect_script" json:"postConnectScript"`
	ConnectScriptIsSudo                 bool               `db:"connect_script_is_sudo" json:"connectScriptIsSudo"`
	StripeWebhookSecret                 *dbcrypto.String   `db:"stripe_webhook_secret"            json:"-"`
	StripeWebhookSecretConfigured       bool               `db:"stripe_webhook_secret_configured" json:"stripeWebhookSecretConfigured"` //nolint:lll
}

func (org *Organization) HasFeature(feature Feature) bool {
	return slices.Contains(org.Features, feature)
}

func (org *Organization) AddFeatures(features ...Feature) {
	for _, feature := range features {
		if !org.HasFeature(feature) {
			org.Features = append(org.Features, feature)
		}
	}
}

func (org *Organization) RemoveFeatures(features ...Feature) {
	org.Features = slices.DeleteFunc(org.Features, func(f Feature) bool { return slices.Contains(features, f) })
}

func (org *Organization) SetFeature(feature Feature, enabled bool) {
	if enabled {
		org.AddFeatures(feature)
	} else {
		org.RemoveFeatures(feature)
	}
}

// SubscriptionPlan is everything an organization's subscription is made of, whatever grants it:
// a Stripe subscription, the license key of a self-hosted instance, or the defaults of a new
// organization.
type SubscriptionPlan struct {
	Type                    SubscriptionType
	Period                  SubscriptionPeriod
	EndsAt                  time.Time
	CustomerOrganizationQty limit.Limit
	UserAccountQty          limit.Limit
}

func (plan SubscriptionPlan) Features() []Feature {
	return FeaturesForSubscriptionType(plan.Type)
}

// ApplyPlan puts the organization on the given plan. The features of the plan are added but none
// are ever removed, so a feature granted outside of a plan survives, and so does a feature of a
// plan the organization is no longer on.
func (org *Organization) ApplyPlan(plan SubscriptionPlan) {
	org.SubscriptionType = plan.Type
	org.SubscriptionPeriod = plan.Period
	org.SubscriptionEndsAt = plan.EndsAt
	org.SubscriptionCustomerOrganizationQty = plan.CustomerOrganizationQty
	org.SubscriptionUserAccountQty = plan.UserAccountQty
	org.AddFeatures(plan.Features()...)
}

func (org *Organization) HasActiveSubscription() bool {
	return org.SubscriptionType == SubscriptionTypeCommunity || org.SubscriptionEndsAt.After(time.Now())
}

func (org *Organization) HasActiveSubscriptionWithType(st SubscriptionType) bool {
	return org.HasActiveSubscription() && org.SubscriptionType == st
}

type OrganizationWithUserRole struct {
	Organization
	UserRole                 UserRole   `db:"user_role" json:"userRole"`
	CustomerOrganizationID   *uuid.UUID `db:"customer_organization_id" json:"customerOrganizationId,omitempty"`
	CustomerOrganizationName *string    `db:"customer_organization_name" json:"customerOrganizationName,omitempty"`
	PartnerOrganizationID    *uuid.UUID `db:"partner_organization_id" json:"partnerOrganizationId,omitempty"`
	PartnerOrganizationName  *string    `db:"partner_organization_name" json:"partnerOrganizationName,omitempty"`
	JoinedOrgAt              time.Time  `db:"joined_org_at" json:"joinedOrgAt"`
}

type OrganizationWithBranding struct {
	Organization
	Branding *OrganizationBranding `db:"branding"`
}

type OrganizationBranding struct {
	ID                     uuid.UUID  `db:"id" json:"id"`
	CreatedAt              time.Time  `db:"created_at" json:"createdAt"`
	OrganizationID         uuid.UUID  `db:"organization_id" json:"-"`
	UpdatedAt              time.Time  `db:"updated_at" json:"updatedAt"`
	UpdatedByUserAccountID *uuid.UUID `db:"updated_by_user_account_id" json:"-"`
	Title                  *string    `db:"title" json:"title"`
	Description            *string    `db:"description" json:"description"`
	LogoImageID            *uuid.UUID `db:"logo_image_id" json:"logoImageId"`
	AppDomain              *string    `db:"app_domain" json:"appDomain"`
	RegistryDomain         *string    `db:"registry_domain" json:"registryDomain"`
	EmailFromAddress       *string    `db:"email_from_address" json:"emailFromAddress"`
	PageTitle              *string    `db:"page_title" json:"pageTitle"`
	FaviconImageID         *uuid.UUID `db:"favicon_image_id" json:"faviconImageId"`
}

File: frontend/ui/src/app/types/organization.ts
import {BaseModel, Named, UserRole} from '@distr-sh/distr-sdk';
import {isExpiredSubscription, SubscriptionType} from './subscription';

export type Feature =
  | 'licensing'
  | 'pre_post_scripts'
  | 'artifact_version_mutable'
  | 'vendor_billing'
  | 'deployment_logs_after'
  | 'partner_management'
  | 'custom_domains'
  | 'vulnerabilities'
  | 'custom_emails'
  | 'custom_oidc_providers';

export interface SubscriptionLimits {
  maxCustomerOrganizations: number;
  maxUsersPerCustomerOrganization: number;
  maxDeploymentsPerCustomerOrganization: number;
  // Reported for display purposes only, this limit is not enforced.
  maxRegistryStorageBytes: number;
  logQueryWindowSeconds: number;
}

export interface CreateUpdateOrganizationRequest {
  name: string;
  slug?: string;
  preConnectScript?: string;
  postConnectScript?: string;
  connectScriptIsSudo: boolean;
  artifactVersionMutable: boolean;
  prePostScriptsEnabled: boolean;
}

export interface Organization extends BaseModel, Named {
  name: string;
  slug?: string;
  features: Feature[];
  subscriptionType: SubscriptionType;
  subscriptionLimits: SubscriptionLimits;
  subscriptionEndsAt?: string;
  subscriptionCustomerOrganizationQuantity: number;
  subscriptionUserAccountQuantity: number;
  currentBillableUserAccountCount: number;
  currentCustomerOrganizationCount: number;
  preConnectScript?: string;
  postConnectScript?: string;
  connectScriptIsSudo: boolean;
  stripeWebhookSecretConfigured: boolean;
}

export interface OrganizationWithUserRole extends Organization {
  userRole: UserRole;
  customerOrganizationId?: string;
  customerOrganizationName?: string;
  partnerOrganizationId?: string;
  partnerOrganizationName?: string;
  joinedOrgAt: string;
}

export function isSubscriptionExpired(org: Organization): boolean {
  return isExpiredSubscription(org.subscriptionType, org.subscriptionEndsAt);
}
